lib/promql/expr.go

```diff
@@ -0,0 +1,423 @@
+}
+
+// ForSubquery returns whether is rollup is for a subquery
+func (re *RollupExpr) ForSubquery() bool {
+	return len(re.Step) > 0 || re.InheritStep
+}
+
+// AppendString appends string representation of expr to dst.
+func (re *RollupExpr) AppendString(dst []byte) []byte {
+	needParens := func() bool {
+		if _, ok := re.Expr.(*RollupExpr); ok {
+			return true
+		}
+		if _, ok := re.Expr.(*BinaryOpExpr); ok {
+			return true
+		}
+		if ae, ok := re.Expr.(*AggrFuncExpr); ok && ae.Modifier.Op != "" {
+			return true
+		}
+		return false
+	}()
+	if needParens {
+		dst = append(dst, '(')
+	}
+	dst = re.Expr.AppendString(dst)
+	if needParens {
+		dst = append(dst, ')')
+	}
+	if len(re.Window) > 0 || re.InheritStep || len(re.Step) > 0 {
+		dst = append(dst, '[')
+		if len(re.Window) > 0 {
+			dst = append(dst, re.Window...)
+		}
+		if len(re.Step) > 0 {
+			dst = append(dst, ':')
+			dst = append(dst, re.Step...)
+		} else if re.InheritStep {
+			dst = append(dst, ':')
+		}
+		dst = append(dst, ']')
+	}
+	if len(re.Offset) > 0 {
+		dst = append(dst, " offset "...)
+		dst = append(dst, re.Offset...)
+	}
+	return dst
+}
+
+// A MetricExpr represents a metric expression
+type MetricExpr struct {
+	// TagFilters contains a list of tag filters from curly braces.
+	// The first item may be the metric name.
+	TagFilters []TagFilter
+}
+
+// AppendString appends string representation of expr to dst.
+func (me *MetricExpr) AppendString(dst []byte) []byte {
+	tfs := me.TagFilters
+	if len(tfs) > 0 {
+		tf := &tfs[0]
+		if len(tf.Key) == 0 && !tf.IsNegative && !tf.IsRegexp {
+			dst = appendEscapedIdent(dst, tf.Value)
+			tfs = tfs[1:]
+		}
+	}
+	if len(tfs) > 0 {
+		dst = append(dst, '{')
+		for i := range tfs {
+			tf := &tfs[i]
+			dst = appendStringTagFilter(dst, tf)
+			if i+1 < len(tfs) {
+				dst = append(dst, ", "...)
+			}
+		}
+		dst = append(dst, '}')
+	} else if len(me.TagFilters) == 0 {
+		dst = append(dst, "{}"...)
+	}
+	return dst
+}
+
+// IsEmpty returns whether this is an empty metric expression
+func (me *MetricExpr) IsEmpty() bool {
+	return len(me.TagFilters) == 0
+}
+
+// IsOnlyMetricGroup returns whether this is a metric group only
+func (me *MetricExpr) IsOnlyMetricGroup() bool {
+	if !me.HasNonEmptyMetricGroup() {
+		return false
+	}
+	return len(me.TagFilters) == 1
+}
+
+// HasNonEmptyMetricGroup returns whether this has a non-empty metric group
+func (me *MetricExpr) HasNonEmptyMetricGroup() bool {
+	if len(me.TagFilters) == 0 {
+		return false
+	}
+	tf := &me.TagFilters[0]
+	return len(tf.Key) == 0 && !tf.IsNegative && !tf.IsRegexp
+}
+
+// A TagFilter is a single key <op> value filter tag in a metric filter
+//
+// Note that this should exactly match the definition in the stroage package
+type TagFilter struct {
+	Key        []byte
+	Value      []byte
+	IsNegative bool
+	IsRegexp   bool
+}
+
+// A MetricTemplateExpr represents a metric expression prior to expansion via
+// a with clause
+type MetricTemplateExpr struct {
```

Comment:
Instead of having a single node that can be in two states (in the past `MetricExpr` had `TagFilters` and `tagFilters`, and depending on the whether the node had been WITH expanded one or the other was populated) I split into a separate node type, and applying the WITH clause changes this to a `MetricExpr`